beater/extension.go

```diff
@@ -0,0 +1,296 @@
+		// we consider this is a continuation of previous line
+		if oldLog, found := jb.journalTypeOutstandingLogBuffer[logType]; found {
+			jb.journalTypeOutstandingLogBuffer[logType].logEvent[messageField] =
+				oldLog.logEvent[messageField].(string) + "\n" + newLogMessage
+		} else {
+			jb.journalTypeOutstandingLogBuffer[logType] = toLogBuffer(event)
+		}
+		jb.journalTypeOutstandingLogBuffer[logType].time = time.Now()
+	} else {
+		oldLogBuffer, found := jb.journalTypeOutstandingLogBuffer[logType]
+		jb.journalTypeOutstandingLogBuffer[logType] = toLogBuffer(event)
+		if found {
+			// flush the older logs to async.
+			jb.publishEvent(oldLogBuffer)
+			// update stats if enabled
+			if jb.config.MetricsEnabled {
+				jb.metrics.logMessagesPublished.Inc(1)
+				jb.metrics.logMessageDelay.Update(time.Now().Unix() - (event[utcTimestampField].(int64) / microseconds))
+			}
+		}
+	}
+}
+
+func (jbe *JournalBeatExtension) publishEvent(logBuffer *LogBuffer) {
+	partition := getPartition(logBuffer, jbe.numLogstashAvailable)
+	jbe.logstashClients[partition].PublishEvent(logBuffer.logEvent, publisher.Guaranteed)
+}
+
+func toLogBuffer(event common.MapStr) *LogBuffer {
+	return &LogBuffer{
+		time:     time.Now(),
+		logType:  event[logBufferingTypeField].(string),
+		logEvent: event,
+	}
+}
+
+// TODO optimize this later but for now walk through all the different types. Use priority queue/multiple threads if needed.
+func (jb *Journalbeat) logProcessor() {
+	logp.Info("Started the thread which consumes log messages and publishes them")
+	tickChan := time.NewTicker(jb.config.FlushLogInterval)
+	for {
+		select {
+		case <-jb.done:
+			return
+		case <-tickChan.C:
+			// here we need to walk through all the map entries and flush out the ones
+			// which have been sitting there for some time.
+			jb.flushStaleEvents()
+
+		case channelEvent := <-jb.incomingLogEvents:
+			jb.flushOrBufferEvent(channelEvent)
+		}
+	}
+}
+
+var commonFields = StringSet{
+	hostNameField:  true,
+	messageField:   true,
+	priorityField:  true,
+	inputTypeField: true,
+}
+
+var containerFields = commonFields.union(StringSet{
+	containerTagField: true,
+	containerIdField:  true,
+})
+
+var nativeFields = commonFields.union(StringSet{
+	tagField:       true,
+	processIdField: true,
+})
+
+func (jbe *JournalBeatExtension) sendEvent(event common.MapStr, rawEvent *sdjournal.JournalEntry) {
+	var newEvent common.MapStr
+	if containerId, exists := event[containerIdField]; exists {
+		newEvent = cloneFields(event, containerFields)
+		newEvent["type"] = "container"
+		newEvent[logBufferingTypeField] = containerId
+	} else if processId, exists := event[processIdField]; exists {
+		newEvent = cloneFields(event, nativeFields)
+		newEvent["type"] = event[tagField]
+		newEvent[logBufferingTypeField] = processId
+	} else {
+		// TODO review. Drop?
+		logp.Info("Strange message %s", event)
+		newEvent["type"] = "unknown"
```

Comment:
Do we want to send these kind of messages downstream or just log it to identify it is a bad un-expected message?